Name: Faris Assallami

In [122]:
## Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
## Importing the OS and JSON Modules
import os,json

In [123]:
df1 = pd.read_csv('/Users/faris/Downloads/superhero_powers - superhero_powers.csv')
df2 = pd.read_csv('/Users/faris/Downloads/superhero_info - superhero_info.csv')
display(df1.info())
display(df2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


None

In [124]:
df1.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [125]:
df2.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [126]:
## adding expand=True
df2['Hero|Publisher'].str.split('|',expand=True)

## save the 2 new columns into the dataframe
df2[['Hero','Publisher']] = df2['Hero|Publisher'].str.split('|',expand=True)

## drop the original column 
df2 = df2.drop(columns=['Hero|Publisher'])
df2.head()


Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [127]:
## examining a single value from the coordinates col
measurements = df2.loc[0,"Measurements"]
print(type(measurements))
measurements

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [128]:
import json
json.loads(measurements)

JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)

In [129]:
measurements

"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [130]:
measurements = measurements.replace("'",'"')
measurements

'{"Height": "203.0 cm", "Weight": "441.0 kg"}'

In [131]:
## now we can use json.loads
fixed_measurements = json.loads(measurements)
print(type(fixed_measurements))
fixed_measurements

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [132]:
## use .str.replace to replace all single quotes
df2['Measurements'] = df2['Measurements'].str.replace("'",'"')
## Apply the json.loads to the full column
df2['Measurements'] = df2['Measurements'].apply(json.loads)
df2['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [133]:
## check a single value after transformation
test_measurements = df2.loc[0, 'Measurements']
print(type(test_measurements))
test_measurements

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [134]:
hei_wei = df2['Measurements'].apply(pd.Series)
hei_wei

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

In [135]:
## use .str.replace to replace all single quotes units in height and weight
hei_wei['Height'] = hei_wei['Height'].str.replace('cm',"")

## Apply the json.loads to the full column
hei_wei['Height'] = hei_wei['Height'].apply(json.loads)
hei_wei['Height'].head()



0    203.0
1    191.0
2    185.0
3    203.0
4    193.0
Name: Height, dtype: float64

In [136]:
hei_wei['Weight'] = hei_wei['Weight'].str.replace('kg',"")

## Apply the json.loads to the full column
hei_wei['Weight'] = hei_wei['Weight'].apply(json.loads)
hei_wei['Weight'].head()

0    441.0
1     65.0
2     90.0
3    441.0
4    122.0
Name: Weight, dtype: float64

In [137]:
# converting the Height and Weight columns to numeric
hei_wei = hei_wei.astype(float)

In [138]:
# Checking to see if type changed to float
hei_wei.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Height  463 non-null    float64
 1   Weight  463 non-null    float64
dtypes: float64(2)
memory usage: 7.4 KB


In [139]:
# concat long_lat with original dataframe
df3 = pd.concat((df2, hei_wei), axis = 1)
df3.drop(columns='Measurements', inplace=True)
display(df3.head())
display(df3.info())

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher  Height  Weight  
0         A-Bomb      Marvel Comics   203.0   441.0  
1     Abe Sapien  Dark Horse Comics   191.0    65.0  
2       Abin Sur          DC Comics   185.0    90.0  
3    Abomination      Marvel Comics   203.0   441.0  
4  Absorbing Man      Marvel Comics   193.0   122.0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Gender      463 non-null    object 
 1   Race        463 non-null    object 
 2   Alignment   463 non-null    object 
 3   Hair color  463 non-null    object 
 4   Eye color   463 non-null    object 
 5   Skin color  463 non-null    object 
 6   Hero        463 non-null    object 
 7   Publisher   463 non-null    object 
 8   Height      463 non-null    float64
 9   Weight      463 non-null    float64
dtypes: float64(2), object(8)
memory usage: 36.3+ KB


None

In [140]:
df1.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [141]:
## showing the lists are really strings
powtyp = df1.loc[7,'Powers']
type(powtyp)

str

In [142]:
# Converting powers column string to lists
df1['powers_split'] = df1['Powers'].str.split(',')
df1.head()



hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        powers_split  
0    [Agility, Super Strength, Stamina, Super Speed]  
1  [Accelerated Healing, Durability, Longevity, S...  
2  [Agility, Accelerated Healing, Cold Resistance...  
3                               [Lantern Power Ring]  
4  [Accelerated Healing, Intelligence, Super Stre...

In [143]:
# exploding the power split column
exploded = df1.explode('powers_split')
exploded.head()

hero_names                                             Powers  \
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   

          powers_split  
0              Agility  
0       Super Strength  
0              Stamina  
0          Super Speed  
1  Accelerated Healing

In [144]:
## saving the unique values from the exploded column
cols_to_make = exploded['powers_split'].dropna().unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [145]:
for col in cols_to_make:
    df1[col] = df1['Powers'].str.contains(col)
df1.head()

/var/folders/15/t06gywxj5nj9956_98qpphjc0000gn/T/ipykernel_81997/779434845.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df1[col] = df1['Powers'].str.contains(col)
/var/folders/15/t06gywxj5nj9956_98qpphjc0000gn/T/ipykernel_81997/779434845.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df1[col] = df1['Powers'].str.contains(col)
/var/folders/15/t06gywxj5nj9956_98qpphjc0000gn/T/ipykernel_81997/779434845.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many ti

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        powers_split  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   
3                               [Lantern Power Ring]    False           False   
4  [Accelerated Healing, Intelligence, Super Stre...    False            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   
3    False        False                False       False      False  ...   
4     True         True                 True       False      False  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 170 columns]

In [146]:
df1 = df1.drop(columns=['Powers','powers_split'])
df1.head()

hero_names  Agility  Super Strength  Stamina  Super Speed  \
0      3-D Man     True            True     True         True   
1       A-Bomb    False            True     True        False   
2   Abe Sapien     True            True     True        False   
3     Abin Sur    False           False    False        False   
4  Abomination    False            True     True         True   

   Accelerated Healing  Durability  Longevity  Camouflage  Self-Sustenance  \
0                False       False      False       False            False   
1                 True        True       True        True             True   
2                 True        True       True       False            False   
3                False       False      False       False            False   
4                 True       False      False       False            False   

   ...  Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0  ...            False        False       False              False   
1  ...            False        False       False              False   
2  ...            False        False       False              False   
3  ...            False        False       False              False   
4  ...            False        False       False              False   

   Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0       False       False          False               False    False   
1       False       False          False               False    False   
2       False       False          False               False    False   
3       False       False          False               False    False   
4       False       False          False               False    False   

   Changing Armor  
0           False  
1           False  
2           False  
3           False  
4           False  

[5 rows x 168 columns]

## Combining both data frames

In [147]:
super_hero_df = pd.merge(df3,df1, left_on= 'Hero', right_on= 'hero_names', how = 'inner')
display(super_hero_df.info())
display(super_hero_df.head())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 463 entries, 0 to 462
Columns: 178 entries, Gender to Changing Armor
dtypes: bool(167), float64(2), object(9)
memory usage: 118.9+ KB


None

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher  Height  Weight  ... Weather Control  \
0         A-Bomb      Marvel Comics   203.0   441.0  ...           False   
1     Abe Sapien  Dark Horse Comics   191.0    65.0  ...           False   
2       Abin Sur          DC Comics   185.0    90.0  ...           False   
3    Abomination      Marvel Comics   203.0   441.0  ...           False   
4  Absorbing Man      Marvel Comics   193.0   122.0  ...           False   

   Omnipresent  Omniscient  Hair Manipulation  Nova Force  Odin Force  \
0        False       False              False       False       False   
1        False       False              False       False       False   
2        False       False              False       False       False   
3        False       False              False       False       False   
4        False       False              False       False       False   

   Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0          False               False    False           False  
1          False               False    False           False  
2          False               False    False           False  
3          False               False    False           False  
4          False               False    False           False  

[5 rows x 178 columns]

In [149]:
# moving Hero column to first column on left side
herocol = super_hero_df.pop('Hero')
super_hero_df.insert(0,'Hero',herocol)

# deleting heros_name column from combined df3
super_hero_df.drop(columns='hero_names')


<class 'pandas.core.frame.DataFrame'>
Int64Index: 463 entries, 0 to 462
Columns: 178 entries, Hero to Changing Armor
dtypes: bool(167), float64(2), object(9)
memory usage: 118.9+ KB


/var/folders/15/t06gywxj5nj9956_98qpphjc0000gn/T/ipykernel_81997/3356003741.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  super_hero_df.insert(0,'Hero',herocol)


None

In [154]:
super_hero_df.head()

Hero Gender               Race Alignment Hair color Eye color  \
0         A-Bomb   Male              Human      good    No Hair    yellow   
1     Abe Sapien   Male      Icthyo Sapien      good    No Hair      blue   
2       Abin Sur   Male            Ungaran      good    No Hair      blue   
3    Abomination   Male  Human / Radiation       bad    No Hair     green   
4  Absorbing Man   Male              Human       bad    No Hair      blue   

  Skin color          Publisher  Height  Weight  ... Weather Control  \
0    Unknown      Marvel Comics   203.0   441.0  ...           False   
1       blue  Dark Horse Comics   191.0    65.0  ...           False   
2        red          DC Comics   185.0    90.0  ...           False   
3    Unknown      Marvel Comics   203.0   441.0  ...           False   
4    Unknown      Marvel Comics   193.0   122.0  ...           False   

   Omnipresent  Omniscient  Hair Manipulation  Nova Force  Odin Force  \
0        False       False              False       False       False   
1        False       False              False       False       False   
2        False       False              False       False       False   
3        False       False              False       False       False   
4        False       False              False       False       False   

   Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0          False               False    False           False  
1          False               False    False           False  
2          False               False    False           False  
3          False               False    False           False  
4          False               False    False           False  

[5 rows x 178 columns]

# 1- Compare the average weight of super powers who have Super Speed to those who do not.

In [160]:
average_height_superspeed = super_hero_df.groupby('Super Speed')['Weight'].mean().round(2)
average_height_superspeed

Super Speed
False    101.77
True     129.40
Name: Weight, dtype: float64

-Average weight of super speed heros is 129.40 kilo

-Average weight of non super speed heros is 101.77 kilo

-Super speed heros on average weigh more than non super speed heros

# 2- What is the average height of heroes for each publisher?

In [158]:
average_height_per_publisher = super_hero_df.groupby('Publisher')['Height'].mean().round(2)
average_height_per_publisher

Publisher
DC Comics            181.92
Dark Horse Comics    176.91
George Lucas         159.60
Image Comics         211.00
Marvel Comics        191.55
Shueisha             171.50
Star Trek            181.50
Team Epic TV         180.75
Unknown              178.00
Name: Height, dtype: float64